bibliopixel/drivers/spi_driver_base.py

```diff
@@ -1,77 +1,168 @@
 
-        # File based SPI requires a bytearray so we have to overwrite _buf
-        if self.use_py_spi:
-            a, b = -1, -1
-            d = self.dev.replace('/dev/spidev', '')
-            s = d.split('.')
-            if len(s) == 2:
-                a = int(s[0])
-                b = int(s[1])
+    def send_packet(self, data):
+        raise NotImplementedError
 
-            if a < 0 or b < 0:
-                error(BAD_FORMAT_ERROR)
+    def compute_packet(self, data):
+        return data
 
-            self._bootstrapSPIDev()
-            import spidev
-            self.spi = spidev.SpiDev()
-            self.spi.open(a, b)
-            self.spi.max_speed_hz = int(self._spiSpeed * 1000000.0)
-            log.info('py-spidev speed @ %.1f MHz',
-                     (float(self.spi.max_speed_hz) / 1000000.0))
-        else:
-            self._buf = bytearray(self._buf)
-            self.spi = self._open(self.dev, 'wb')
-
-    def _bootstrapSPIDev(self):
-        if self.use_py_spi:
-            try:
-                import spidev
-            except:
-                error(CANT_IMPORT_SPIDEV_ERROR)
-
-        if not os.path.exists(self.dev):
+
+# partial source @see https://armbedded.taskit.de/node/318
+# @see <linux/spi/spidev.h>
+# hex numbers from good old strace(1)
+IOC_READ = 0x40000000
+IOC_WRITE = 0x40000000
+
+IOC_BITS_PER_WORD = 0x00016b03
+IOC_MAX_SPEED_HZ = 0x00046b04
+
+
+class SpiFileInterface(SpiBaseInterface):
+    """ using os open/write to send data"""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        if not os.path.exists(self._dev):
             error(CANT_FIND_ERROR)
 
+        try:
+            import fcntl
+            self._fcntl = fcntl
+        except ImportError:
+            error(CANT_IMPORT_FCNTL_ERROR)
+        self._spi = open(self._dev, 'wb')
+
+        self.speed = self._spi_speed
+        self.bits_per_word = 8
+        log.info('file io spi speed @ %.1f MHz, %d bits per word', self.speed / 1e6, self.bits_per_word)
+
+    def _ioctl(self, command, buffer):
+        self._fcntl.ioctl(self._spi, IOC_WRITE + command, buffer)
+        self._fcntl.ioctl(self._spi, IOC_READ + command, buffer)
+
+    def _set_speed(self, speed):
+        buffer = struct.pack(">I", int(speed * 1e9))  # unint32
+        self._ioctl(IOC_MAX_SPEED_HZ, buffer)
+        self._speed = struct.unpack(">I", buffer)[0] / 1000
+
+    def _get_speed(self):
+        return self._speed
+
+    speed = property(_get_speed, _set_speed)
+
+    def _set_bits_per_word(self, bits):
+        buffer = struct.pack("B", bits)  # uint8
+        self._ioctl(IOC_BITS_PER_WORD, buffer)
+        self._bits_per_word = int.from_bytes(buffer, byteorder='big')
+
+    def _get_bits_per_word(self):
+        return self._bits_per_word
+
+    bits_per_word = property(_get_bits_per_word, _set_bits_per_word)
+
+    def send_packet(self, data):
+        self._set_speed(self._spi_speed)  # set speed again if other program changed it
```

Comment:
As I mentioned a time ago, I'm going to change to a other syscall that sets the  speed and all other settings in one command (syscall), but need a bit more time.
Also want not a longer pull request and split it into two (or more), so you can work on the other part of the project and I have a bit time to think about it and test it.
Fhe file io driver happend accidentally, so the pull request get a bit of a mess.

I remove for now that code and also all public function for the settings in the interface.
They were currently only access able in the `DriverSpiBase` and in it's childs. Will bring this back with a common interface for file io and spidev driver and also forward it to the code that uses this lib.

Not sure wenn I find time probably in 2-3 days.